src-ghc/Pact/Types/Crypto.hs

```diff
@@ -15,117 +17,168 @@
 -- PPK and hashing types.
 --
 module Pact.Types.Crypto
-  ( PublicKey, importPublic, exportPublic
-  , PrivateKey, importPrivate, exportPrivate
-  , Signature(..), exportSignature
-  , sign, valid
-  , PPKScheme(..)
-  , Hash(..), initialHash, hashLengthAsBS, hashLengthAsBase16, hashToB16Text
+  ( hashTx, verifyHashTx, initialHashTx
+  , PPKScheme(..), SPPKScheme, SomeScheme(..)
+  , InitScheme(..), Scheme(..), defaultScheme, formatPublicKey
+  , KeyPair(..), SomeKeyPair(..), importKeyPair
+  , PublicKeyText(..), PrivateKeyText(..), SignatureText(..)
+  , hashLengthAsBS, hashLengthAsBase16, hashToB16Text
   ) where
 
+
 import Control.Applicative
 import Control.Monad.Reader
 import Control.DeepSeq
 
-import Crypto.Ed25519.Pure ( PublicKey, PrivateKey, Signature(..), importPublic, importPrivate, exportPublic, exportPrivate)
+import "crypto-api" Crypto.Random
 import qualified Crypto.Ed25519.Pure as Ed25519
+import qualified Crypto.Hash as H
 
 import Data.Aeson as A
 import Data.Aeson.Types (toJSONKeyText)
 
 import Data.ByteString (ByteString)
+import qualified Data.ByteArray as B
 import Data.Serialize as SZ
 import qualified Data.Serialize as S
 import Data.Maybe
 import Data.Text.Encoding
+import qualified Data.Text as T
+
 
 import GHC.Generics
 import Prelude
 
 import Pact.Types.Util
-import Pact.Types.Hash
-
-deriving instance Eq Signature
-deriving instance Ord Signature
-instance Serialize Signature where
-  put (Sig s) = S.put s
-  get = Sig <$> (S.get >>= S.getByteString)
-instance ToJSON Signature where
-  toJSON (Sig s) = toB16JSON s
-instance FromJSON Signature where
-  parseJSON = withText "Signature" parseText
-  {-# INLINE parseJSON #-}
-instance ParseText Signature where
-  parseText s = Sig <$> parseB16Text s
-  {-# INLINE parseText #-}
 
-exportSignature :: Signature -> ByteString
-exportSignature (Sig s) = s
+
+--------- HASHING ---------
+
+hashTx :: (H.HashAlgorithm a) => a -> ByteString -> Hash
+hashTx algo b = (Hash . B.convert . H.hashWith algo) b
+
+verifyHashTx :: (H.HashAlgorithm a) => a -> Hash -> ByteString -> Either String Hash
+verifyHashTx algo h b = if hashTx algo b == h
+  then Right h
+  else Left $ "Hash Mismatch, received " ++ show h
+       ++ " but our hashing resulted in " ++ show (hashTx algo b)
+{-# INLINE verifyHashTx #-}
+
+initialHashTx :: (H.HashAlgorithm a) => a -> Hash
+initialHashTx algo = hashTx algo mempty
+
+
+
+
+--------- SCHEME ---------
 
 data PPKScheme = ED25519
   deriving (Show, Eq, Ord, Generic)
 
+
+-- Run-time witness to PPKScheme kind
+data SPPKScheme (a :: PPKScheme) = MakeScheme
+data SomeScheme = SED25519 (SPPKScheme 'ED25519)
+
+
+
+
+class InitScheme s where
+  toScheme :: s -> SomeScheme
+
+
+
+class (Export (PublicKey a), Export (Signature a),
```

Comment:
At a high level this looks like a pretty reasonable start to me.  I'd really like to see at least one more hash scheme implemented before passing judgment to see how things fall out.

Reply:
Like adding another PPKScheme? i.e. `PPKScheme = ED25519 | ETH`?